src/utils.rs

```diff
@@ -35,6 +37,215 @@ use crate::FrameLayout;
+    }
+
+    /// Whether the stream still has RBSP data. Implements more_rbsp_data(). See
+    /// the spec for more details.
+    pub fn has_more_rsbp_data(&mut self) -> bool {
+        if self.num_remaining_bits_in_curr_byte == 0 && self.update_curr_byte().is_err() {
+            // no more data at all in the rbsp
+            return false;
+        }
+
+        // If the next bit is the stop bit, then we should only see unset bits
+        // until the end of the data.
+        if (self.curr_byte & ((1 << (self.num_remaining_bits_in_curr_byte - 1)) - 1)) != 0 {
+            return true;
+        }
+
+        let data = self.data.chunk();
+        for data in &data[0..self.data.remaining()] {
+            if *data != 0 {
+                return true;
+            }
+        }
+
+        self.data.advance(self.data.remaining());
+        false
+    }
+
+    pub fn read_ue<U: TryFrom<u32>>(&mut self) -> anyhow::Result<U> {
+        let mut num_bits = 0;
+
+        while self.read_bits::<u32>(1)? == 0 {
+            num_bits += 1;
+        }
+
+        if num_bits > 31 {
+            return Err(anyhow!("invalid stream"));
+        }
+
+        let value = ((1u32 << num_bits) - 1)
+            .checked_add(self.read_bits::<u32>(num_bits)?)
+            .context("read number cannot fit in 32 bits")?;
+
+        U::try_from(value).map_err(|_| anyhow!("conversion error"))
+    }
+
+    pub fn read_ue_bounded<U: TryFrom<u32>>(&mut self, min: u32, max: u32) -> anyhow::Result<U> {
+        let ue = self.read_ue()?;
+        if ue > max || ue < min {
+            Err(anyhow!(
+                "Value out of bounds: expected {} - {}, got {}",
+                min,
+                max,
+                ue
+            ))
+        } else {
+            Ok(U::try_from(ue).map_err(|_| anyhow!("Conversion error"))?)
+        }
+    }
+
+    pub fn read_ue_max<U: TryFrom<u32>>(&mut self, max: u32) -> anyhow::Result<U> {
+        self.read_ue_bounded(0, max)
+    }
+
+    pub fn read_se<U: TryFrom<i32>>(&mut self) -> anyhow::Result<U> {
+        let ue = self.read_ue::<u32>()? as i32;
+
+        if ue % 2 == 0 {
+            Ok(U::try_from(-ue / 2).map_err(|_| anyhow!("Conversion error"))?)
+        } else {
+            Ok(U::try_from(ue / 2 + 1).map_err(|_| anyhow!("Conversion error"))?)
+        }
+    }
+
+    pub fn read_se_bounded<U: TryFrom<i32>>(&mut self, min: i32, max: i32) -> anyhow::Result<U> {
+        let se = self.read_se()?;
+        if se < min || se > max {
+            Err(anyhow!(
+                "Value out of bounds, expected between {}-{}, got {}",
+                min,
+                max,
+                se
+            ))
+        } else {
+            Ok(U::try_from(se).map_err(|_| anyhow!("Conversion error"))?)
+        }
+    }
+
+    fn get_byte(&mut self) -> Result<u8, GetByteError> {
+        self.data.read_u8().map_err(|_| GetByteError::OutOfBits)
+    }
+
+    fn update_curr_byte(&mut self) -> Result<(), GetByteError> {
```

Comment:
Optional: I think this private method name could be more self-descriptive.  Maybe "move_to_next_byte"?